crates/ursa-gateway/src/worker/cache.rs

```diff
@@ -69,30 +93,80 @@ impl WorkerCache for Cache {
 
 #[async_trait]
 pub trait ServerCache: Send + Sync + 'static {
-    async fn get_announce(&self, k: &str) -> Result<Arc<Vec<u8>>>;
+    async fn get_announce(&self, k: &str) -> Result<StreamBody<ReaderStream<DuplexStream>>>;
 }
 
 #[async_trait]
 impl ServerCache for Cache {
-    async fn get_announce(&self, k: &str) -> Result<Arc<Vec<u8>>> {
+    async fn get_announce(&self, k: &str) -> Result<StreamBody<ReaderStream<DuplexStream>>> {
+        let (mut w, r) = duplex(self.stream_buf as usize);
         if let Some(data) = self.tlrfu.dirty_get(&String::from(k)) {
-            self.tx.send(WorkerCacheCommand::GetSync {
-                key: String::from(k),
-            })?;
-            Ok(Arc::clone(data))
+            let data = Arc::clone(data);
+            self.tx
+                .send(WorkerCacheCommand::GetSync {
+                    key: String::from(k),
+                })
+                .map_err(|e| {
+                    error!("Failed to dispatch GetSync command: {e:?}");
+                    anyhow!("Failed to dispatch GetSync command")
+                })?;
+            spawn(async move {
+                if let Err(e) = w.write_all(data.as_ref()).await {
+                    error!("Failed to write to stream: {e:?}");
+                }
+            });
         } else {
             let (tx, rx) = oneshot::channel();
-            self.tx.send(WorkerCacheCommand::Fetch {
-                cid: String::from(k),
-                sender: tx,
-            })?;
-            let data = rx.await??;
-            self.tx.send(WorkerCacheCommand::InsertSync {
-                key: String::from(k),
-                value: Arc::clone(&data),
-            })?;
-            Ok(data)
+            self.tx
+                .send(WorkerCacheCommand::Fetch {
+                    cid: String::from(k),
+                    sender: tx,
+                })
+                .map_err(|e| {
+                    error!("Failed to dispatch Fetch command: {e:?}");
+                    anyhow!("Failed to dispatch Fetch command")
+                })?;
+            let mut body = match rx.await??.into_parts() {
+                (
+                    Parts {
+                        status: StatusCode::OK,
+                        ..
+                    },
+                    body,
+                ) => body,
+                resp => {
+                    error!("Error requested provider: {resp:?}");
+                    bail!("Error requested provider");
+                }
+            };
+            let key = String::from(k); // move to [worker|writer] thread
+            let tx = self.tx.clone(); // move to [worker|writer] thread
+            spawn(async move {
```

Comment:
Isn't this still buffering the response from the cache node in memory?

Reply:
This is a streaming feature though so nvm. We can discuss further if we need piping.

Reply:
yup, it's a streaming feature; we might need a pipeline solution one day.